### PR TITLE
add include option to /search command

### PR DESCRIPTION
**Pull request by PCSwingle**
Allows including with /search. Let me know what everyone thinks about the messages; they feel a bit long and clunky to me, but not sure how to shorten them.

### CODE REVIEW

File: mentat/command/commands/search.py [hunk 1]
```diff
@@ -21,19 +23,57 @@ async def apply(self, *args: str) -> None:
...
+        cumulative_tokens = 0
         for i, (feature, score) in enumerate(results, start=1):
+            i_str = f"{i:3}"
+            score_str = f"{score:.3f}"
+
             label = feature.ref()
-            if label.startswith(str(session_context.cwd)):
-                label = label[len(str(session_context.cwd)) + 1 :]
+            label = label.removeprefix(str(session_context.cwd) + "/")
             if feature.name:
```
Comment (PCSwingle) [inline mentat/command/commands/search.py:33]: I personally dislike having the feature.name tacked onto the end of the results; on the one hand, I understand why it's useful (especially when including to know what you're including), but it's also way too long, dense, and unreadable. I think we should either clean it up or remove it. Try using /search with and without it to see what I mean.

**Comment by biobootloader** [comment]
I typed `21` even though only 20 results were shown and it let me, probably should re-ask for invalid inputs?

Also it got several "Interval out of bounds" errors - maybe the last line numbers are 1 higher than file lengths?

image

**Comment by PCSwingle** [comment]
> I typed `21` even though only 20 results were shown and it let me, probably should re-ask for invalid inputs?

This was on purpose; did you only want to allow numbers up to what it's shown so far? We have all of the results. I can definitely change it though if you still want me to.

> Also it got several "Interval out of bounds" errors - maybe the last line numbers are 1 higher than file lengths?

Just looked into this and it's a bug from the git PR's (we check the length of the file by how many lines readlines gives us, but readlines gives the newline at the end of each line and doesn't start a new line if it's empty, so it's off by one sometimes.) I'll fix this bug as well.

**Comment by biobootloader** [comment]
For the formatting, we definitely need to simplify. See screenshot at bottom for how lines overflow even on a wide terminal window.

A few ideas:
- we could drop the scores - is it useful to know the match was 0.75 vs 0.7?
- Add a title row at the top so we don't have the words "Tokens" and "Cumulative Tokens" repeated every line
- fixed width spacing of columns
- A column for the filename, perhaps dropping parts of the full path? So instead of `mentat/feature_filters/llm_feature_filter.py` we could just display `llm_feature_filter.py` or `m/f/llm_feature_filter.py`?
- drop the line numbers, as they'd only be useful if you're comparing with the exact file, which would take time. This is meant to be something you can do really quickly.
- remove quotes around feature name

The feature name is the trickiest part. Hopefully the above ideas give us more room to include it:

> I personally dislike having the feature.name tacked onto the end of the results; on the one hand, I understand why it's useful (especially when including to know what you're including), but it's also way too long, dense, and unreadable. I think we should either clean it up or remove it. Try using /search with and without it to see what I mean.

You are right that it's long and dense, but it's also the only way to know what parts of files are being included, so I think we need it. Perhaps we truncate it if it would cause the line to wrap (though then we need to be able to check the user's terminal size). Or we could wrap more intelligently, like just have it wrap within it's column at the end?

**Original Output:**
```
  1 | 0.756 | 240 Tokens | 240 Cumulative Tokens | mentat/llm_api_handler.py:257-280 "LlmApiHandler.call_embedding_api,LlmApiHandler.is_model_available,LlmApiHandler.call_whisper_api"
  2 | 0.740 | 229 Tokens | 469 Cumulative Tokens | docs/configuration.md:54-69 "🔧 Configuration.🦙 Alternative Models"
  3 | 0.735 | 19 Tokens | 488 Cumulative Tokens | .flake8
  4 | 0.735 | 263 Tokens | 751 Cumulative Tokens | mentat/llm_api_handler.py:181-204 "LlmApiHandler,LlmApiHandler.initialize_client"
  5 | 0.735 | 276 Tokens | 1027 Cumulative Tokens | README.md:126-139 "👩‍💻 Roadmap and Contributing"
  6 | 0.735 | 396 Tokens | 1423 Cumulative Tokens | mentat/cost_tracker.py:11-48 "CostTracker.total_tokens,CostTracker.total_cost,CostTracker.log_api_call_stats"
  7 | 0.735 | 61 Tokens | 1484 Cumulative Tokens | pyrightconfig.json
  8 | 0.734 | 352 Tokens | 1836 Cumulative Tokens | mentat/llm_api_handler.py:222-257 "LlmApiHandler.call_llm_api"
  9 | 0.730 | 165 Tokens | 2001 Cumulative Tokens | mentat/llm_api_handler.py:204-222 "LlmApiHandler.call_llm_api,LlmApiHandler.call_llm_api"
 10 | 0.729 | 172 Tokens | 2173 Cumulative Tokens | README.md:61-73 "⚙️ Setup.Add your OpenAI API Key,⚙️ Setup.Configuration"
 
Enter (Y) for more results, (n) to exit search mode, or a number to add all features above that number to context.
```

 **Reformatted output (thanks GPT-4):**
 ```  | Tokens/Total | Filename           | Feature Name
--+--------------+--------------------+------------------------------------------
1 | 240/240      | llm_api_handler.py | LlmApiHandler.call_embedding_api,LlmApiHandler.is_model_available,LlmApiHandler.call_whisper_api
2 | 229/469      | configuration.md   | 🔧 Configuration.🦙 Alternative Models
3 | 19/488       | .flake8            | 
4 | 263/751      | llm_api_handler.py | LlmApiHandler,LlmApiHandler.initialize_client
5 | 276/1027     | README.md          | 👩‍💻 Roadmap and Contributing
6 | 396/1423     | cost_tracker.py    | CostTracker.total_tokens,CostTracker.total_cost,CostTracker.log_api_call_stats
7 | 61/1484      | pyrightconfig.json | 
8 | 352/1836     | llm_api_handler.py | LlmApiHandler.call_llm_api
9 | 165/2001     | llm_api_handler.py | LlmApiHandler.call_llm_api,LlmApiHandler.call_llm_api
10| 172/2173     | README.md          | ⚙️ Setup.Add your OpenAI API Key,⚙️ Setup.Configuration

Enter (Y) for more results, (n) to exit search mode, or a number to add all features above that number to context.
```

**Reformated version with feature name wrapping in column:**
```
  | Tokens/Total | Filename           | Feature Name
--+--------------+--------------------+------------------------------------------
1 | 240/240      | llm_api_handler.py | LlmApiHandler.call_embedding_api,
   |              |                    | LlmApiHandler.is_model_available,
   |              |                    | LlmApiHandler.call_whisper_api
2 | 229/469      | configuration.md   | 🔧 Configuration.🦙 Alternative Models
3 | 19/488       | .flake8            | 
4 | 263/751      | llm_api_handler.py | LlmApiHandler,
   |              |                    | LlmApiHandler.initialize_client
5 | 276/1027     | README.md          | 👩‍💻 Roadmap and Contributing
6 | 396/1423     | cost_tracker.py    | CostTracker.total_tokens,
   |              |                    | CostTracker.total_cost,
   |              |                    | CostTracker.log_api_call_stats
7 | 61/1484      | pyrightconfig.json | 
8 | 352/1836     | llm_api_handler.py | LlmApiHandler.call_llm_api
9 | 165/2001     | llm_api_handler.py | LlmApiHandler.call_llm_api,
   |              |                    | LlmApiHandler.call_llm_api
10| 172/2173     | README.md          | ⚙️ Setup.Add your OpenAI API Key,
   |              |                    | ⚙️ Setup.Configuration

Enter (Y) for more results, (n) to exit search mode, or a number to add all features above that number to context.
```

image

File: mentat/command/commands/search.py (hunk 1 above)
Comment (jakethekoenig) [reply mentat/command/commands/search.py:33]: I agree

**Comment by jakethekoenig** [comment]
I like full paths. If you're searching for something adding it to context to help mentat is only one use case. You might also want to open it in your own editor to read the code. For that reason I also like keeping line numbers.

I do like:
* dropping the feature name
* Making titles
* Omitting cumulative tokens: it overindexes on the one use case and even in that case people can eyeball the sum.
* Omitting cosine similarity

**Comment by PCSwingle** [comment]
New idea:
```
1 | 240/240 Tokens/Total
  | mentat/llm_api_handler.py:10-30
  | LlmApiHandler.call_embedding_api, 
  | LlmApiHandler.is_model_available, 
  | LlmApiHandler.call_whisper_api
2 | 240/480 Tokens/Total
  | mentat/llm_api_handler.py:30-50
  | LlmApiHandler.call_embedding_api, 
  | LlmApiHandler.is_model_available, 
  | LlmApiHandler.call_whisper_api
3 | 240/720 Tokens/Total
  | mentat/llm_api_handler.py:62:87
  | LlmApiHandler.call_embedding_api, 
  | LlmApiHandler.is_model_available, 
  | LlmApiHandler.call_whisper_api
```
@jakethekoenig what do you think about this?

**Comment by jakethekoenig** [comment]
> New idea:
> 
> ```
> 1 | 240/240 Tokens/Total
>   | mentat/llm_api_handler.py:10-30
>   | LlmApiHandler.call_embedding_api, 
>   | LlmApiHandler.is_model_available, 
>   | LlmApiHandler.call_whisper_api
>   |
> 2 | 240/480 Tokens/Total
>   | mentat/llm_api_handler.py:30-50
>   | LlmApiHandler.call_embedding_api, 
>   | LlmApiHandler.is_model_available, 
>   | LlmApiHandler.call_whisper_api
>   |
> 3 | 240/720 Tokens/Total
>   | mentat/llm_api_handler.py:62:87
>   | LlmApiHandler.call_embedding_api, 
>   | LlmApiHandler.is_model_available, 
>   | LlmApiHandler.call_whisper_api
> ```
> 
> @jakethekoenig what do you think about this?

At first I found it confusing. But it's growing on me. I think it needs some indentation. I'd still prefer to omit cumulative information.

**Comment by biobootloader** [comment]
> You might also want to open it in your own editor to read the code. For that reason I also like keeping line numbers.

good point

> Omitting cumulative tokens: it overindexes on the one use case and even in that case people can eyeball the sum.

I guess that makes sense, lets try it without cumulative tokens

> dropping the feature name

I really think we need feature names - that's the only way to quickly know what's actually included

**Comment by biobootloader** [comment]
How about we move filepath to top line now that we are dropping cumulative tokens? Something like this:

```
1 | 240 tokens | mentat/llm_api_handler.py:10-30
  |   LlmApiHandler.call_embedding_api, 
  |   LlmApiHandler.is_model_available, 
  |   LlmApiHandler.call_whisper_api
  |
2 | 300 tokens | mentat/llm_api_handler.py:30-50
  |    LlmApiHandler.call_embedding_api, 
  |    LlmApiHandler.is_model_available, 
  |    LlmApiHandler.call_whisper_api
  |
```

**Comment by PCSwingle** [comment]
Ok here's the current iteration; let me know what you think. Haven't dropped cumulative tokens yet but open to the idea:
image

**Comment by PCSwingle** [comment]
Next iteration:
image

**Comment by jakethekoenig** [comment]
ooh, I like the colors. They make everything better.

Part of why I think count/cumulative is confusing is because it makes it look like a fraction. Like I see `312/407` and my first thought it "Oh no this file uses up most of ... something". I think 5 is fine. It's easy to scroll and with precise searches the correct stuff seems to be often at the top.

**Comment by PCSwingle** [comment]
> ooh, I like the colors. They make everything better.
> 
> Part of why I think count/cumulative is confusing is because it makes it look like a fraction. Like I see `312/407` and my first thought it "Oh no this file uses up most of ... something". I think 5 is fine. It's easy to scroll and with precise searches the correct stuff seems to be often at the top.

Hmm I think I agree with you. I'll remove the total tokens and then push and everybody can test.

**Comment by jakethekoenig** [comment]
For the record I actually think the right UX would be the user can enter a list of numbers and have them added to context. As is rhere will be annoying cases where for whatever reason the Readme or some piece of documentation is above what you actually want to add. But I suppose there's always `/include` to fall back on.

**Comment by PCSwingle** [comment]
> For the record I actually think the right UX would be the user can enter a list of numbers and have them added to context. As is rhere will be annoying cases where for whatever reason the Readme or some piece of documentation is above what you actually want to add. But I suppose there's always `/include` to fall back on.

I actually agree with this; a couple of ways (other than yours, which would also work well) would be to give every resulta 
 symbol (starting with 1, so it might be 123456789abcdefg...) and we would have enough symbols for more than enough results and to include a specific result you just type that symbol. Or we could do it more like pacman's include packages to install prompt where it auto includes everything (in this case it would auto include everything it's shown the user) and you can type in the specific numbers (space separated) to exclude them (or 1-4 to exclude all from 1-4, ^5 to only include 5, etc.). What do you think @biobootloader ?

**Comment by granawkins** [comment]
FWIW I'd expect to see a list 1-10 and then say '2,6' and have features 2 and 6 added to context.

**Comment by biobootloader** [comment]
> For the record I actually think the right UX would be the user can enter a list of numbers and have them added to context.

> FWIW I'd expect to see a list 1-10 and then say '2,6' and have features 2 and 6 added to context.

That does feel like what most people will expect...

> As is there will be annoying cases where for whatever reason the Readme or some piece of documentation is above what you actually want to add.

my thinking is that this will be used to be fast and dirty: instead of trying to precisely add specific sections, you just want to get the related stuff added quickly without worrying about a few extra things making their way in. But yeah maybe that's not how most people will see it.

> 123456789abcdefg

It seems awkward if the user has shown more results a couple times, they'll have 30 labeled results to pick from. If they are individually picking probably they need to pick results from a section before they move to the next section? Otherwise they'd be scrolling up to see which one was 3 or 4, c or d, etc.

**Comment by biobootloader** [comment]
Maybe the flow could be like this?

- /search "my search query"
- see the first 10 results. Type in `1 2 5 6 9` to choose those
- it'd re-display the same list, so you can add another or choose to quit the search or move to next section
- Type `y` to see next section
- see next results 10 results, type `5 6 8` to add a few
- type `n` to exit search

Maybe on a section if you typed `all` it would add everything in that section?

**Comment by PCSwingle** [comment]
> Maybe the flow could be like this?
> 
> * /search "my search query"
> * see the first 10 results. Type in `1 2 5 6 9` to choose those
> * it'd re-display the same list, so you can add another or choose to quit the search or move to next section
> * Type `y` to see next section
> * see next results 10 results, type `5 6 8` to add a few
> * type `n` to exit search
> 
> Maybe on a section if you typed `all` it would add everything in that section?

I think this makes sense to me. I also think we should support 1-3 and commas in the addition format. This would also mean we don't need 'all' (since user can just do 1-5). I don't like 'all' too much because we would have to add that to the message and it feels weird that it would only add everything in a section instead of all the shown results.

**Comment by PCSwingle** [comment]
Inspired by yay, I really like how they do it (although in our case it's reversed; include instead of exclude)
image

**Comment by PCSwingle** [comment]
Just changed it so you can accept specific results; let me know what everyone thinks! There are a couple of things I'm iffy on, but if nobody else dislikes them then I'll assume it's good.

**Comment by biobootloader** [comment]
Nice!

One small thing: after you exist let's put a gap and the usual prompt ("What can I do for you?"):

image

File: mentat/command/commands/search.py [hunk 2]
```diff
@@ -26,19 +49,60 @@ async def apply(self, *args: str) -> None:
             stream.send(str(e), color="red")
             return
 
-        for i, (feature, score) in enumerate(results, start=1):
-            label = feature.ref()
-            if label.startswith(str(session_context.cwd)):
-                label = label[len(str(session_context.cwd)) + 1 :]
+        cumulative_tokens = 0
```
Comment (biobootloader) [inline mentat/command/commands/search.py:52]: no longer used

**Comment by biobootloader** [comment]
I think it looks really good except for maybe the placement of "396 tokens". Maybe it should be on the first line in parenthesis after the filename / lines? Also I think a batch size of 10 is better, and with the removed tokens line it's more compact so easier to increase batch size.

# Current
image


# Potential Changes
image

File: mentat/command/commands/search.py [hunk 3]
```diff
@@ -26,19 +49,58 @@ async def apply(self, *args: str) -> None:
...
                 # TODO: Required to avoid circular imports, but not ideal.
-                from mentat.session_input import ask_yes_no
+                from mentat.session_input import collect_user_input
 
-                stream.send("\nShow More results? ")
-                if not await ask_yes_no(default_yes=True):
+                stream.send(
+                    "(Y/n) for more results or to exit search mode.\nResults to"
+                    ' include in context: (eg: "1 3 4" or "1-4")'
+                )
```
Comment (biobootloader) [inline mentat/command/commands/search.py:82]: is this more clear?

```suggestion
                stream.send(
                    "'Y' to show more results\n"
                    "'n' to exit search mode\n"
                    "or select specific results to add to context (e.g., '1 3 4' or '1-4'):"
                )
```

**Review by biobootloader — APPROVED** [review]
left a couple comments but looks good!

File: mentat/command/commands/search.py (hunk 3 above)
Comment (PCSwingle) [reply mentat/command/commands/search.py:82]: I think I prefer the current one; more concise.

**Pull request merged**